fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/query/plan/synthetic/SyntheticRecordPlannerTest.java

```diff
@@ -1040,6 +1049,90 @@ public void aggregateJoinIndex() throws Exception {
+        );
+        joined.addJoin("order",
+                field("custRef").nest("string_value"),
+                "cust",
+                function(IntWrappingFunction.NAME, field("___header").nest("int_rec_id"))
+        );
+
+        metaDataBuilder.addIndex(joined, new Index("joinNestedConcat", concat(
+                field("cust").nest("name"),
+                field("order").nest("order_no")
+        )));
+
+        // Add index on custRef field to facilitate finding join partners of customer records
+        metaDataBuilder.addIndex("OrderWithHeader", new Index("order$custRef", concat(
+                field("___header").nest("z_key"),
+                field("custRef").nest("string_value")
+        )));
+
+        try (FDBRecordContext context = openContext()) {
+            final FDBRecordStore recordStore = recordStoreBuilder.setContext(context).create();
+            final Index joinIndex = recordStore.getRecordMetaData().getIndex("joinNestedConcat");
+            recordStore.markIndexDisabled(joinIndex).get();
+            context.commit();
+        }
+
+        IndexMaintainerRegistry backingRegistry = IndexMaintainerRegistryImpl.instance();
+        IndexMaintainerRegistry registry = new IndexMaintainerRegistry() {
+            @Nonnull
+            @Override
+            public IndexMaintainer getIndexMaintainer(@Nonnull final IndexMaintainerState state) {
+                if(!state.index.getName().equals("joinNestedConcat")){
+                    return backingRegistry.getIndexMaintainer(state);
+                }
+                else return new FailIndexMaintainer(state);
+            }
+
+            @Nonnull
+            @Override
+            public IndexValidator getIndexValidator(@Nonnull final Index index) {
+                return backingRegistry.getIndexValidator(index);
+            }
+        };
+
+        try (FDBRecordContext context = openContext()) {
+            final FDBRecordStore recordStore = recordStoreBuilder
+                    .setIndexMaintainerRegistry(registry)
+                    .setContext(context).open();
+
+            TestRecordsJoinIndexProto.CustomerWithHeader customer = TestRecordsJoinIndexProto.CustomerWithHeader.newBuilder()
+                    .setHeader(TestRecordsJoinIndexProto.Header.newBuilder().setZKey(1).setIntRecId(1L))
+                    .setName("Scott")
+                    .setCity("Toronto")
+                    .build();
+            recordStore.saveRecord(customer);
+
+            TestRecordsJoinIndexProto.OrderWithHeader order = TestRecordsJoinIndexProto.OrderWithHeader.newBuilder()
+                    .setHeader(TestRecordsJoinIndexProto.Header.newBuilder().setZKey(1).setRecId("23"))
+                    .setOrderNo(10)
+                    .setQuantity(23)
+                    .setCustRef(TestRecordsJoinIndexProto.Ref.newBuilder().setStringValue("i:1"))
+                    .build();
+            recordStore.saveRecord(order);
+        }
+    }
```

Comment:
This test seems like it mostly tests what we want, in that the test should fail if the buggy maintainer gets called. The other way I can think to test this would be to create a normal join index, disable it, then insert some records that would generate entries if the index wasn't disabled, and then call `uncheckedMarkIndexReadable` to force the index into a readable state without issuing a build. Then, the index should be scannable (and the test can validate no entries were written).

The other thing that we may want to test: I could be wrong, but I _think_ the current approach will not even try and find the join constituents on a record if the index(es) it would be used for are `DISABLED`. If we wanted to test that, then I think we could do that in one of two ways:

1. Create a join that the `SyntheticRecordPlanner` can't correctly calculate joins for. For example, using non-invertible `FunctionKeyExpression` in the join constituent might do it
2. Modify the `SyntheticRecordPlanner` so that it will return plans that throw errors when run. That's a little more difficult, given how the `SyntheticRecordPlanner` is constructed, though maybe adding in enough hooks is possible. An alternative alternative to that might be to add metrics surrounding the amount of time being spent loading synthetic records (which are metrics we may want anyway), and then asserting that these metrics are 0 if the index is DISABLED

Reply:
the current structure of the code requires the underlying RL planner to be able to create a plan for both sides of the join _before_ it determines if the indexes are enabled or not. This is arguably the incorrect order--we probably want to not even attempt to plan the join elements if all the destination indexes are disabled. But that is how it currently stands.

Reply:
But the plans don't get executed in that case, right? So both my proposals to either have plans that throw errors when run or asserting that timer values associated with running the plans are zero would be consistent with the current code structure, I'd think